BadImageFormatException on codegeneration in 2.0.0-beta1, targetting netstandard2.0 win7-x64

Hi,

When trying to build a grain implementations project, targetting netstandard2.0, runtime win7-x64 I get the following exception:

``` 
5>Orleans-CodeGen - Generating file D:\Projects\AO\src\api\Ao.Core.Grains\obj\Debug\netstandard2.0\win7-x64\Ao.Core.Grains.orleans.g.cs
5>warn: Orleans.CodeGenerator.SerializerGenerationManager[103807]
5>      System type ClaimsPrincipal may require a custom serializer for optimal performance. If you use arguments of this type a lot, consider submitting a pull request to https://github.com/dotnet/orleans/ to add a custom serializer for it.
5>-- Code Generation FAILED --
5>
5>Exc level 0: System.BadImageFormatException: Could not load file or assembly 'System.Data.SqlClient, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Reference assemblies should not be loaded for execution.  They can only be loaded in the Reflection-only loader context. (Exception from HRESULT: 0x80131058)
5>   at System.Signature.GetSignature(Void* pCorSig, Int32 cCorSig, RuntimeFieldHandleInternal fieldHandle, IRuntimeMethodInfo methodHandle, RuntimeType declaringType)
5>   at System.Reflection.RuntimeConstructorInfo.get_Signature()
5>   at System.Reflection.RuntimeConstructorInfo.GetParametersNoCopy()
5>   at System.RuntimeType.FilterApplyMethodBase(MethodBase methodBase, BindingFlags methodFlags, BindingFlags bindingFlags, CallingConventions callConv, Type[] argumentTypes)
5>   at System.RuntimeType.GetConstructorCandidates(String name, BindingFlags bindingAttr, CallingConventions callConv, Type[] types, Boolean allowPrefixLookup)
5>   at System.RuntimeType.GetConstructorImpl(BindingFlags bindingAttr, Binder binder, CallingConventions callConvention, Type[] types, ParameterModifier[] modifiers)
5>   at Orleans.CodeGenerator.SerializerGenerator.GetObjectCreationExpressionSyntax(Type type)
5>   at Orleans.CodeGenerator.SerializerGenerator.GenerateDeepCopierMethod(Type type, List`1 fields)
5>   at Orleans.CodeGenerator.SerializerGenerator.GenerateClass(String className, Type type, Action`1 onEncounteredType)
5>   at Orleans.CodeGenerator.RoslynCodeGenerator.GenerateForAssemblies(Assembly targetAssembly, Boolean runtime)
5>   at Orleans.CodeGenerator.RoslynCodeGenerator.GenerateSourceForAssembly(Assembly input)
5>   at Orleans.CodeGeneration.CodeGenerator.GenerateSourceForAssembly(Assembly grainAssembly)
5>   at Orleans.CodeGeneration.CodeGenerator.GenerateCodeInternal(CodeGenOptions options)
5>   at Orleans.CodeGeneration.CodeGenerator.GenerateCode(CodeGenOptions options)
5>   at Orleans.CodeGeneration.Program.Main(String[] args)
5>Exc level 1: System.BadImageFormatException: Cannot load a reference assembly for execution.
5>C:\Users\r.bosch\.nuget\packages\microsoft.orleans.orleanscodegenerator.build\2.0.0-beta1\build\Microsoft.Orleans.OrleansCodeGenerator.Build.targets(69,5): error MSB3073: The command ""dotnet" "C:\Users\r.bosch\.nuget\packages\microsoft.orleans.orleanscodegenerator.build\2.0.0-beta1\build\..\tasks\netcoreapp2.0\Orleans.CodeGeneration.Build.dll" "@obj\Debug\netstandard2.0\win7-x64\codegen\Ao.Core.Grains.orleans.g.args.txt"" exited with code 3.
5>Done building project "Ao.Core.Grains.csproj" -- FAILED.
========== Build: 4 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
``` 

targetting net462, runtime win7-x64 it succeeds, with warnings:

```
5>Orleans-CodeGen - Generating file D:\Projectens\AO\src\api\Ao.Core.Grains\obj\Debug\net462\win7-x64\Ao.Core.Grains.orleans.g.cs
5>warn: Orleans.CodeGenerator.SerializerGenerationManager[103807]
5>      System type ClaimsPrincipal may require a custom serializer for optimal performance. If you use arguments of this type a lot, consider submitting a pull request to https://github.com/dotnet/orleans/ to add a custom serializer for it.
5>warn: Orleans.CodeGenerator.RoslynCodeGenerator[101723]
5>      Exception loading types from assembly 'Core.Services, Version=0.1.0.0, Culture=neutral, PublicKeyToken=null':
5>      Exc level 0: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
5>         at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
5>         at System.Reflection.RuntimeAssembly.get_DefinedTypes()
5>         at Orleans.Runtime.TypeUtils.GetDefinedTypes(Assembly assembly, Logger logger) in D:\build\agent\_work\18\s\src\Orleans.Core\CodeGeneration\TypeUtils.cs:line 555
5>      Exc level 1: System.BadImageFormatException: Could not load file or assembly 'System.Data.SqlClient, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. Reference assemblies should not be loaded for execution.  They can only be loaded in the Reflection-only loader context. (Exception from HRESULT: 0x80131058)
5>      Exc level 2: System.BadImageFormatException: Cannot load a reference assembly for execution..
5>System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
5>   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
5>   at System.Reflection.RuntimeAssembly.get_DefinedTypes()
5>   at Orleans.Runtime.TypeUtils.GetDefinedTypes(Assembly assembly, Logger logger) in D:\build\agent\_work\18\s\src\Orleans.Core\CodeGeneration\TypeUtils.cs:line 555
5>Orleans-CodeGen - Generated file written D:\Projectens\AO\src\api\Ao.Core.Grains\obj\Debug\net462\win7-x64\Ao.Core.Grains.orleans.g.cs
5>Ao.Core.Grains -> D:\Projectens\AO\src\api\Ao.Core.Grains\bin\Debug\net462\win7-x64\Ao.Core.Grains.dll
========== Build: 5 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

--- a/src/Orleans.CodeGeneration.Build/AssemblyResolver.cs
+++ b/src/Orleans.CodeGeneration.Build/AssemblyResolver.cs
@@ -22,7 +22,8 @@ namespace Orleans.CodeGeneration
         /// Needs to be public so can be serialized accross the the app domain.
         /// </summary>
         public Dictionary<string, string> ReferenceAssemblyPaths { get; } = new Dictionary<string, string>();
-        
+
+        private readonly bool usedByCodeGenerator;
         private readonly ICompilationAssemblyResolver assemblyResolver;
 
         private readonly DependencyContext dependencyContext;
@@ -31,10 +32,14 @@ namespace Orleans.CodeGeneration
         private readonly AssemblyLoadContext loadContext;
 #endif
 
-        public AssemblyResolver(string path, List<string> referencedAssemblies)
+        public AssemblyResolver(string path, List<string> referencedAssemblies, bool usedByCodeGenerator = false)
         {
-            if (Path.GetFileName(path) == "Orleans.Core.dll")  this.Assembly = typeof(RuntimeVersion).Assembly;
-            else this.Assembly = Assembly.LoadFrom(path);
+            this.usedByCodeGenerator = usedByCodeGenerator;
+
+            if (Path.GetFileName(path) == "Orleans.Core.dll")
+                this.Assembly = typeof(RuntimeVersion).Assembly;
+            else
+                this.Assembly = Assembly.LoadFrom(path);
 
             this.dependencyContext = DependencyContext.Load(this.Assembly);
             this.resolverRependencyContext = DependencyContext.Load(typeof(AssemblyResolver).Assembly);
@@ -48,16 +53,27 @@ namespace Orleans.CodeGeneration
                     new PackageCompilationAssemblyResolver()
                 });
 
-            AppDomain.CurrentDomain.AssemblyResolve += this.ResolveAssembly;
 #if NETCOREAPP2_0
             this.loadContext = AssemblyLoadContext.GetLoadContext(this.Assembly);
-            this.loadContext.Resolving += this.AssemblyLoadContextResolving;
-            if (this.loadContext != AssemblyLoadContext.Default)
+
+            if (this.loadContext == AssemblyLoadContext.Default)
             {
-                AssemblyLoadContext.Default.Resolving += this.AssemblyLoadContextResolving;
+                if (!this.usedByCodeGenerator)
+                {
+                    AssemblyLoadContext.Default.Resolving += this.AssemblyLoadContextResolving;
+                }
+            }
+            else
+            {
+                this.loadContext.Resolving += this.AssemblyLoadContextResolving;
+            }
+#else
+            if (!this.usedByCodeGenerator)
+            {
+                AppDomain.CurrentDomain.AssemblyResolve += this.ResolveAssembly;
             }
 #endif
-            
+
             foreach (var assemblyPath in referencedAssemblies)
             {
                 var libName = Path.GetFileNameWithoutExtension(assemblyPath);
@@ -71,13 +87,23 @@ namespace Orleans.CodeGeneration
 
         public void Dispose()
         {
-            AppDomain.CurrentDomain.AssemblyResolve -= this.ResolveAssembly;
-
 #if NETCOREAPP2_0
-            this.loadContext.Resolving -= this.AssemblyLoadContextResolving;
-            if (this.loadContext != AssemblyLoadContext.Default)
+
+            if (this.loadContext == AssemblyLoadContext.Default)
             {
-                AssemblyLoadContext.Default.Resolving -= this.AssemblyLoadContextResolving;
+                if (!this.usedByCodeGenerator)
+                {
+                    AssemblyLoadContext.Default.Resolving -= this.AssemblyLoadContextResolving;
+                }
+            }
+            else
+            {
+                this.loadContext.Resolving -= this.AssemblyLoadContextResolving;
+            }
+#else
+            if (!this.usedByCodeGenerator)
+            {
+                AppDomain.CurrentDomain.AssemblyResolve -= this.ResolveAssembly;
             }
 #endif
         }
@@ -92,15 +118,16 @@ namespace Orleans.CodeGeneration
         /// </returns>
         public Assembly ResolveAssembly(object sender, ResolveEventArgs args)
         {
+            var context = default(AssemblyLoadContext);
+
 #if NETCOREAPP2_0
-            var context = AssemblyLoadContext.GetLoadContext(args.RequestingAssembly);
-#else
-            AssemblyLoadContext context = null;
+            context = AssemblyLoadContext.GetLoadContext(args.RequestingAssembly);
 #endif
+
             return this.AssemblyLoadContextResolving(context, new AssemblyName(args.Name));
         }
 
-        public Assembly AssemblyLoadContextResolving(AssemblyLoadContext context, AssemblyName name)
+            public Assembly AssemblyLoadContextResolving(AssemblyLoadContext context, AssemblyName name)
         {
             // Attempt to resolve the library from one of the dependency contexts.
             var library = this.resolverRependencyContext?.RuntimeLibraries?.FirstOrDefault(NamesMatch)
@@ -127,20 +154,27 @@ namespace Orleans.CodeGeneration
                 }
             }
 
-            if (this.ReferenceAssemblyPaths.TryGetValue(name.FullName, out string path))
+            if (this.ReferenceAssemblyPaths.TryGetValue(name.FullName, out var pathByFullName))
             {
-                var assembly = TryLoadAssemblyFromPath(path);
+                var assembly = TryLoadAssemblyFromPath(pathByFullName);
                 if (assembly != null) return assembly;
             }
 
-            if (this.ReferenceAssemblyPaths.TryGetValue(name.Name, out path))
+            if (this.ReferenceAssemblyPaths.TryGetValue(name.Name, out var pathByName))
             {
-                var assembly = TryLoadAssemblyFromPath(path);
-                if (assembly != null) return assembly;
+                //
+                // Only try to load it if the resolved path is different than from before
+                //
+
+                if (String.Compare(pathByFullName, pathByName, StringComparison.OrdinalIgnoreCase) != 0)
+                {
+                    var assembly = TryLoadAssemblyFromPath(pathByName);
+                    if (assembly != null) return assembly;
+                }
             }
-                
+
             return null;
-            
+
             bool NamesMatch(RuntimeLibrary runtime)
             {
                 return string.Equals(runtime.Name, name.Name, StringComparison.OrdinalIgnoreCase);

--- a/src/Orleans.CodeGeneration.Build/AssemblyResolver.cs
+++ b/src/Orleans.CodeGeneration.Build/AssemblyResolver.cs
@@ -23,7 +23,7 @@ namespace Orleans.CodeGeneration
         /// </summary>
         public Dictionary<string, string> ReferenceAssemblyPaths { get; } = new Dictionary<string, string>();
 
-        private readonly bool usedByCodeGenerator;
+        private readonly bool installDefaultResolveHandler;
         private readonly ICompilationAssemblyResolver assemblyResolver;
 
         private readonly DependencyContext dependencyContext;
@@ -32,9 +32,9 @@ namespace Orleans.CodeGeneration
         private readonly AssemblyLoadContext loadContext;
 #endif
 
-        public AssemblyResolver(string path, List<string> referencedAssemblies, bool usedByCodeGenerator = false)
+        public AssemblyResolver(string path, List<string> referencedAssemblies, bool installDefaultResolveHandler = true)
         {
-            this.usedByCodeGenerator = usedByCodeGenerator;
+            this.installDefaultResolveHandler = installDefaultResolveHandler;
 
             if (Path.GetFileName(path) == "Orleans.Core.dll")
                 this.Assembly = typeof(RuntimeVersion).Assembly;
@@ -58,7 +58,7 @@ namespace Orleans.CodeGeneration
 
             if (this.loadContext == AssemblyLoadContext.Default)
             {
-                if (!this.usedByCodeGenerator)
+                if (this.installDefaultResolveHandler)
                 {
                     AssemblyLoadContext.Default.Resolving += this.AssemblyLoadContextResolving;
                 }
@@ -68,7 +68,7 @@ namespace Orleans.CodeGeneration
                 this.loadContext.Resolving += this.AssemblyLoadContextResolving;
             }
 #else
-            if (!this.usedByCodeGenerator)
+            if (this.installDefaultResolveHandler)
             {
                 AppDomain.CurrentDomain.AssemblyResolve += this.ResolveAssembly;
             }
@@ -91,7 +91,7 @@ namespace Orleans.CodeGeneration
 
             if (this.loadContext == AssemblyLoadContext.Default)
             {
-                if (!this.usedByCodeGenerator)
+                if (this.installDefaultResolveHandler)
                 {
                     AssemblyLoadContext.Default.Resolving -= this.AssemblyLoadContextResolving;
                 }
@@ -101,7 +101,7 @@ namespace Orleans.CodeGeneration
                 this.loadContext.Resolving -= this.AssemblyLoadContextResolving;
             }
 #else
-            if (!this.usedByCodeGenerator)
+            if (this.installDefaultResolveHandler)
             {
                 AppDomain.CurrentDomain.AssemblyResolve -= this.ResolveAssembly;
             }

--- a/src/Orleans.CodeGeneration.Build/CodeGenerator.cs
+++ b/src/Orleans.CodeGeneration.Build/CodeGenerator.cs
@@ -158,7 +158,7 @@ namespace Orleans.CodeGeneration
             var referencedAssemblies = options.ReferencedAssemblies;
 
             // Set up assembly resolver
-            var refResolver = new AssemblyResolver(inputAssembly, referencedAssemblies, true);
+            var refResolver = new AssemblyResolver(inputAssembly, referencedAssemblies, false);
 
             try
             {

--- a/src/Orleans.CodeGeneration.Build/CodeGenerator.cs
+++ b/src/Orleans.CodeGeneration.Build/CodeGenerator.cs
@@ -158,14 +158,15 @@ namespace Orleans.CodeGeneration
             var referencedAssemblies = options.ReferencedAssemblies;
 
             // Set up assembly resolver
-            var refResolver = new AssemblyResolver(inputAssembly, referencedAssemblies);
+            var refResolver = new AssemblyResolver(inputAssembly, referencedAssemblies, true);
 
             try
             {
                 // Set up assembly resolution.
-                AppDomain.CurrentDomain.AssemblyResolve += refResolver.ResolveAssembly;
 #if NETCOREAPP2_0
                 AssemblyLoadContext.Default.Resolving += refResolver.AssemblyLoadContextResolving;
+#else
+                AppDomain.CurrentDomain.AssemblyResolve += refResolver.ResolveAssembly;
 #endif
 
                 return GenerateSourceForAssembly(refResolver.Assembly);
@@ -173,9 +174,10 @@ namespace Orleans.CodeGeneration
             finally
             {
                 refResolver.Dispose();
-                AppDomain.CurrentDomain.AssemblyResolve -= refResolver.ResolveAssembly;
 #if NETCOREAPP2_0
                 AssemblyLoadContext.Default.Resolving -= refResolver.AssemblyLoadContextResolving;
+#else
+                AppDomain.CurrentDomain.AssemblyResolve -= refResolver.ResolveAssembly;
 #endif
             }
         }

--- a/src/Orleans.CodeGeneration/SerializerGenerator.cs
+++ b/src/Orleans.CodeGeneration/SerializerGenerator.cs
@@ -420,7 +420,7 @@ namespace Orleans.CodeGenerator
                 // Use the default value.
                 result = SF.DefaultExpression(typeInfo.AsType().GetTypeSyntax());
             }
-            else if (typeInfo.GetConstructor(Type.EmptyTypes) != null)
+            else if (GetEmptyConstructor(typeInfo) != null)
             {
                 // Use the default constructor.
                 result = SF.ObjectCreationExpression(typeInfo.AsType().GetTypeSyntax()).AddArgumentListArguments();
@@ -436,6 +436,36 @@ namespace Orleans.CodeGenerator
                     getUninitializedObject.Invoke()
                         .AddArgumentListArguments(
                             SF.Argument(SF.TypeOfExpression(typeInfo.AsType().GetTypeSyntax()))));
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Return a parameterless ctor if found. Since typeInfo.GetConstructor can throw BadImageFormatException, 
+        /// we have to do the manual filtering here.
+        /// </summary>
+        /// <param name="typeInfo">The typeInfo</param>
+        /// <returns>Given ctor or null if not found.</returns>
+        private static ConstructorInfo GetEmptyConstructor(TypeInfo typeInfo)
+        {
+            var result = default(ConstructorInfo);
+            var ctors = typeInfo.GetConstructors();
+
+            foreach (var ctor in ctors)
+            {
+                try
+                {
+                    if (ctor.GetParameters().Length==0)
+                    {
+                        result = ctor;
+                        break;
+                    }
+                }
+                catch (BadImageFormatException)
+                {
+                    // Intentionally left blank
+                }
             }
 
             return result;

--- a/src/Orleans.Core/CodeGeneration/TypeUtils.cs
+++ b/src/Orleans.Core/CodeGeneration/TypeUtils.cs
@@ -569,9 +569,7 @@ namespace Orleans.Runtime
 
                         if (logger != null && logger.IsWarning)
                         {
-                            var badImageFormatExceptionCount = typeLoadException.LoaderExceptions.Count(ex => ex as BadImageFormatException != null);
-
-                            if (typeLoadException.LoaderExceptions.Count() != badImageFormatExceptionCount)
+                            if (typeLoadException.LoaderExceptions.Any(ex => !(ex is BadImageFormatException)) || logger.IsVerbose)
                             {
                                 var message =
                                     $"Exception loading types from assembly '{assembly.FullName}': {LogFormatter.PrintException(exception)}.";

--- a/src/Orleans.Core/CodeGeneration/TypeUtils.cs
+++ b/src/Orleans.Core/CodeGeneration/TypeUtils.cs
@@ -174,7 +174,7 @@ namespace Orleans.Runtime
             {
                 return GetParameterizedTemplateName(GetSimpleTypeName(typeInfo, fullName), typeInfo, applyRecursively, fullName);
             }
-            
+
             var t = typeInfo.AsType();
             if (fullName != null && fullName(t) == true)
             {
@@ -190,7 +190,7 @@ namespace Orleans.Runtime
                 fullName = tt => false;
 
             if (!typeInfo.IsGenericType) return baseName;
-            
+
             string s = baseName;
             s += "<";
             bool first = true;
@@ -401,7 +401,7 @@ namespace Orleans.Runtime
             }
 
             if (grainType == type || grainChevronType == type) return false;
-            
+
             if (!grainType.IsAssignableFrom(type)) return false;
 
             // exclude generated classes.
@@ -504,7 +504,7 @@ namespace Orleans.Runtime
 
             return generalType.IsAssignableFrom(type) && TypeHasAttribute(type, typeof(MethodInvokerAttribute));
         }
-        
+
 
         private static readonly Lazy<bool> canUseReflectionOnly = new Lazy<bool>(() =>
         {
@@ -556,24 +556,45 @@ namespace Orleans.Runtime
             }
             catch (Exception exception)
             {
+                var typeLoadException = exception as ReflectionTypeLoadException;
+
+                if (typeLoadException != null)
+                {
+                    if (typeLoadException.LoaderExceptions != null)
+                    {
+                        //
+                        // If we've only BadImageFormatExceptions in LoaderExceptions, then it's ok to not to log, otherwise log
+                        // as a warning.
+                        //
+
+                        if (logger != null && logger.IsWarning)
+                        {
+                            var badImageFormatExceptionCount = typeLoadException.LoaderExceptions.Count(ex => ex as BadImageFormatException != null);
+
+                            if (typeLoadException.LoaderExceptions.Count() != badImageFormatExceptionCount)
+                            {
+                                var message =
+                                    $"Exception loading types from assembly '{assembly.FullName}': {LogFormatter.PrintException(exception)}.";
+                                logger.Warn(ErrorCode.Loader_TypeLoadError_5, message, exception);
+                            }
+                        }
+                    }
+
+                    return typeLoadException.Types?.Where(type => type != null).Select(type => type.GetTypeInfo()) ??
+                           Enumerable.Empty<TypeInfo>();
+                }
+
                 if (logger != null && logger.IsWarning)
                 {
                     var message =
                         $"Exception loading types from assembly '{assembly.FullName}': {LogFormatter.PrintException(exception)}.";
                     logger.Warn(ErrorCode.Loader_TypeLoadError_5, message, exception);
                 }
-                
-                var typeLoadException = exception as ReflectionTypeLoadException;
-                if (typeLoadException != null)
-                {
-                    return typeLoadException.Types?.Where(type => type != null).Select(type => type.GetTypeInfo()) ??
-                           Enumerable.Empty<TypeInfo>();
-                }
 
                 return Enumerable.Empty<TypeInfo>();
             }
         }
-        
+
         /// <summary>
         /// Returns a value indicating whether or not the provided <paramref name="methodInfo"/> is a grain method.
         /// </summary>
@@ -879,7 +900,7 @@ namespace Orleans.Runtime
 
             throw new ArgumentException("Expression type unsupported.");
         }
-        
+
         /// <summary>
         /// Returns the <see cref="PropertyInfo"/> for the simple member access in the provided <paramref name="expression"/>.
         /// </summary>

